lib/ruby_lsp/requests/support/dependency_detector.rb

```diff
@@ -0,0 +1,41 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  module DependencyDetector
+    class << self
+      extend T::Sig
+
+      sig { returns(String) }
+      def detected_formatter
+        # NOTE: Intentionally no $ at end, since we want to match rubocop-shopify, etc.
+        if direct_dependency?(/^rubocop/)
+          "rubocop"
+        elsif direct_dependency?(/^syntax_tree$/)
+          "syntax_tree"
+        else
+          "none"
+        end
+      end
+
+      sig { returns(String) }
+      def detected_test_library
```

Comment:
With the current implementation, this will run on every code lens request - even though the test framework wouldn't change.

How about using `Singleton` here to evaluate the formatter and test framework ahead of time?